virttest/cartesian_config.py

```diff
@@ -538,10 +538,64 @@ def dump(self, indent, recurse=False):
 match_substitute = re.compile("\$\{(.+?)\}")
 
 
+def _drop_suffixes(d):
+    """
+    Merge suffixes for same var, or drop off unnecessary suffixes
+
+    This step returns a copy of a suffix flattened dictionary.
+    """
+    # dictionary `d_flat' is going to be the mutated copy of `d`
+    d_flat = d.copy()
+    for key in d:
+        if key in _reserved_keys:
+            continue
+
+        if not isinstance(key, tuple):
+            continue
+
+        try:
+            # This file was invoked through cmdline
+            options.skipdups
+            skipdups = options.skipdups
+        except NameError:
+            # This file was invoked as Python module
+            skipdups = True
+
+        if skipdups:
+            # Drop vars with suffixes matches general var val
+            # Example: if a_x == 1, and a == 1. Drop: a_x, leave a
+            gen_var_name = key[0]
+            if gen_var_name in d and d[gen_var_name] == d[key]:
+                # Drop gen_var_name, use general key with same value
+                d_flat.pop(key)
+                continue
+
+            can_drop_all_suffixes_for_this_key = True
+            for k in d:
```

Comment:
Maybe in the future it is possible to refactor this function to iterate over the dict twice in separate `for` loops (e.g. by grouping tuples with the same prefix in a second `dict` in the first pass and handling then separately in a second iteration). Since this function is called multiple times, it would be interesting to reduce its time complexity.

Reply:
Very good idea, however indeed a bit too risky and out of scope for the current PR. I will note it down though.